File: src/moniker_svc/telemetry/events.py
"""Telemetry event types."""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Any


class EventOutcome(str, Enum):
    """Outcome of a moniker access."""
    SUCCESS = "success"
    NOT_FOUND = "not_found"
    ERROR = "error"
    UNAUTHORIZED = "unauthorized"
    RATE_LIMITED = "rate_limited"


class Operation(str, Enum):
    """Type of operation performed."""
    READ = "read"
    LIST = "list"
    DESCRIBE = "describe"
    LINEAGE = "lineage"
    REQUEST_SUBMIT = "request_submit"
    REQUEST_APPROVE = "request_approve"
    REQUEST_REJECT = "request_reject"
    REQUEST_COMMENT = "request_comment"


@dataclass(frozen=True, slots=True)
class CallerIdentity:
    """
    Identity of the caller making the request.

    At minimum, one of these should be present.
    """
    # Service identity (from mTLS or service account)
    service_id: str | None = None

    # User identity (from OAuth/JWT)
    user_id: str | None = None

    # Application/client identifier
    app_id: str | None = None

    # Team/department (for chargeback)
    team: str | None = None

    # Additional claims from auth token
    claims: dict[str, Any] = field(default_factory=dict)

    @property
    def principal(self) -> str:
        """Primary identifier for this caller."""
        return self.service_id or self.user_id or self.app_id or "anonymous"

    def __str__(self) -> str:
        return self.principal


@dataclass(frozen=True, slots=True)
class UsageEvent:
    """
    A single usage event for telemetry.

    Captures everything needed for:
    - Data lineage ("what data was accessed")
    - Governance ("who accessed what")
    - Chargeback ("which team/app used how much")
    - Debugging ("what went wrong")
    """
    # Request identification
    request_id: str

    # Timing
    timestamp: datetime

    # Who
    caller: CallerIdentity

    # What
    moniker: str  # Full moniker string
    moniker_domain: str | None  # Domain segment
    moniker_path: str  # Path without params
    operation: Operation

    # Outcome
    outcome: EventOutcome
    error_message: str | None = None

    # Performance
    latency_ms: float = 0.0

    # Resolution details
    resolved_source_type: str | None = None
    resolved_source_path: str | None = None

    # Ownership at access time (for audit)
    owner_at_access: str | None = None

    # Row/record count if applicable
    result_count: int | None = None

    # Was this served from cache?
    cached: bool = False

    # Additional context
    metadata: dict[str, Any] = field(default_factory=dict)

    # Deprecation telemetry
    deprecated: bool = False
    successor: str | None = None
    redirected_from: str | None = None

    @classmethod
    def create(
        cls,
        moniker: str,
        moniker_path: str,
        operation: Operation,
        caller: CallerIdentity,
        outcome: EventOutcome,
        latency_ms: float = 0.0,
        **kwargs,
    ) -> UsageEvent:
        """Factory method with sensible defaults."""
        # Extract domain from path
        domain = moniker_path.split("/")[0] if "/" in moniker_path else moniker_path

        return cls(
            request_id=str(uuid.uuid4()),
            timestamp=datetime.now(timezone.utc),
            caller=caller,
            moniker=moniker,
            moniker_domain=domain,
            moniker_path=moniker_path,
            operation=operation,
            outcome=outcome,
            latency_ms=latency_ms,
            **kwargs,
        )

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary for serialization."""
        return {
            "request_id": self.request_id,
            "timestamp": self.timestamp.isoformat(),
            "caller": {
                "principal": self.caller.principal,
                "service_id": self.caller.service_id,
                "user_id": self.caller.user_id,
                "app_id": self.caller.app_id,
                "team": self.caller.team,
            },
            "moniker": self.moniker,
            "moniker_domain": self.moniker_domain,
            "moniker_path": self.moniker_path,
            "operation": self.operation.value,
            "outcome": self.outcome.value,
            "error_message": self.error_message,
            "latency_ms": self.latency_ms,
            "resolved_source_type": self.resolved_source_type,
            "resolved_source_path": self.resolved_source_path,
            "owner_at_access": self.owner_at_access,
            "result_count": self.result_count,
            "cached": self.cached,
            "metadata": self.metadata,
            "deprecated": self.deprecated,
            "successor": self.successor,
            "redirected_from": self.redirected_from,
        }

File: resolver-go/internal/telemetry/events.go
package telemetry

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
)

// EventOutcome represents the result of a request
type EventOutcome string

const (
	OutcomeSuccess      EventOutcome = "success"
	OutcomeNotFound     EventOutcome = "not_found"
	OutcomeError        EventOutcome = "error"
	OutcomeUnauthorized EventOutcome = "unauthorized"
	OutcomeRateLimited  EventOutcome = "rate_limited"
)

// Operation represents the type of operation performed
type Operation string

const (
	OperationRead    Operation = "read"
	OperationList    Operation = "list"
	OperationDescribe Operation = "describe"
	OperationLineage Operation = "lineage"
)

// CallerIdentity represents the identity of the API caller
type CallerIdentity struct {
	ServiceID *string                `json:"service_id,omitempty"`
	UserID    *string                `json:"user_id,omitempty"`
	AppID     *string                `json:"app_id,omitempty"`
	Team      *string                `json:"team,omitempty"`
	Claims    map[string]interface{} `json:"claims,omitempty"`
}

// Principal returns a human-readable principal identifier
func (c *CallerIdentity) Principal() string {
	if c.ServiceID != nil && *c.ServiceID != "" {
		return fmt.Sprintf("service:%s", *c.ServiceID)
	}
	if c.UserID != nil && *c.UserID != "" {
		return fmt.Sprintf("user:%s", *c.UserID)
	}
	if c.AppID != nil && *c.AppID != "" {
		return fmt.Sprintf("app:%s", *c.AppID)
	}
	return "anonymous"
}

// UsageEvent represents a single telemetry event
type UsageEvent struct {
	// Request identification
	RequestID string    `json:"request_id"`
	Timestamp time.Time `json:"timestamp"`

	// Caller identity
	Caller CallerIdentity `json:"caller"`

	// Moniker details
	Moniker     string `json:"moniker"`
	MonikerPath string `json:"moniker_path"`

	// Operation details
	Operation Operation      `json:"operation"`
	Outcome   EventOutcome   `json:"outcome"`
	LatencyMS float64        `json:"latency_ms"`

	// Resolution details
	ResolvedSourceType *string `json:"resolved_source_type,omitempty"`
	OwnerAtAccess      *string `json:"owner_at_access,omitempty"`

	// Deprecation tracking
	Deprecated     bool    `json:"deprecated"`
	Successor      *string `json:"successor,omitempty"`
	RedirectedFrom *string `json:"redirected_from,omitempty"`

	// Error information
	ErrorMessage *string `json:"error_message,omitempty"`

	// Performance
	Cached bool `json:"cached"`

	// Additional metadata
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}

// NewUsageEvent creates a new usage event with defaults
func NewUsageEvent(moniker, path string, caller CallerIdentity, operation Operation) *UsageEvent {
	return &UsageEvent{
		RequestID:   uuid.New().String(),
		Timestamp:   time.Now().UTC(),
		Caller:      caller,
		Moniker:     moniker,
		MonikerPath: path,
		Operation:   operation,
		Outcome:     OutcomeSuccess,
		Deprecated:  false,
		Cached:      false,
		Metadata:    make(map[string]interface{}),
	}
}

// MarshalJSON customizes JSON marshaling to ensure timestamp format
func (e *UsageEvent) MarshalJSON() ([]byte, error) {
	type Alias UsageEvent
	return json.Marshal(&struct {
		Timestamp string `json:"timestamp"`
		*Alias
	}{
		Timestamp: e.Timestamp.Format(time.RFC3339Nano),
		Alias:     (*Alias)(e),
	})
}

// CompactString returns a one-line compact representation of the event
func (e *UsageEvent) CompactString() string {
	outcome := string(e.Outcome)
	if e.ErrorMessage != nil {
		outcome = fmt.Sprintf("%s: %s", outcome, *e.ErrorMessage)
	}

	return fmt.Sprintf("[TELEMETRY] %s %s %s %s %s %.1fms",
		e.Timestamp.Format(time.RFC3339),
		e.Caller.Principal(),
		e.Operation,
		e.MonikerPath,
		outcome,
		e.LatencyMS,
	)
}
